File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.js
function solution(diffs, times, limit) {
    let left = 1, right = 100001;

    function solveTime(level) {
        let total = 0;

        for (let i = 0; i < diffs.length; i++) {
            const diff = diffs[i];
            const cur = times[i];

            if (diff <= level) total += cur;
            else {
                const mistakes = diff - level;
                const prev = times[i-1]
                
                total += mistakes * (prev + cur) + cur
            }
        }

        return total;
    }

    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        
        if (solveTime(mid) <= limit) right = mid - 1;
        else left = mid + 1;
    }

    return left;
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    left, right = 1, max(diffs)
    answer = right
    
    def solveTime(level):
        total = 0
        prev = times[0]
        
        for i in range(len(diffs)):
            diff = diffs[i]
            cur = times[i]
            
            if diff <= level: total += cur
            else:
                mistakes = diff - level
                total += (mistakes * (cur + prev)) + cur
            prev = cur
        
        return total
    
    while left <= right:
        mid = (left + right) // 2
        if solveTime(mid) <= limit:
            answer = mid
            right = mid - 1
        else: left = mid + 1
    
    return answer
